util/svdgen.py

```diff
@@ -0,0 +1,604 @@
+# Copyright lowRISC contributors.
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+"""
+Generate SVD file from validated register JSON tree
+"""
+
+import hjson
+import pathlib
+import pysvd
+import subprocess
+import sys
+import xml.etree.ElementTree as ET
+
+
+GENERATED_NOTICE = '''
+This file generated from HJSON source by "svdgen.py", do not edit.'''
+
+def read_git_version() -> str:
+    """Read the repository version string from Git"""
+
+    describe = 'git describe --tags --long --dirty'.split()
+    describe = subprocess.run(describe, stdout=subprocess.PIPE)
+    if describe.returncode != 0:
+        raise SystemExit('failed to determine Git repository version: %s' %
+                str(describe.stderr, 'UTF-8'))
+
+    return str(describe.stdout, 'UTF-8').strip()
+
+def read_relative_path(*components: [str]) -> str:
+    """Read the contents of a file relative to the current source."""
+
+    source_dir = pathlib.Path(__file__).parent
+    return source_dir.joinpath(*components).read_text()
+
+def value(num: int) -> str or None:
+    """Converts None -> None and everything else to hex"""
+
+    if num != None:
+        return hex(num)
+
+def ordinal(n: int) -> str:
+    """Convert an integer into its English ordinal number: 1st, 2nd, ...
+    The returned value is 1-based to account for sequence numbering in a
+    list (intended for human consumption)."""
+
+    tens, digit = divmod(n%100, 10)
+    suffix = 'stndrdth'[2*min(digit + (tens==1)*3, 3):]
+    return '%i%s' % (n+1, suffix[:2])
+
+def create(element: str or ET.Element, *elements: [ET.Element], **texts: {"tag": object}) -> ET.Element:
```

Comment:
I've changed the name to `svd_node` to more accurately describe the functionality.

I'm torn on mixing *args and kwargs. I agree that it can lend some confusion—the out-of-order generation in `generate_device()` is a perfect example. (I tried a few ways to avoid this, but couldn't convince the interpreter to make it work.) Additionally, use of kwargs results in random ordering of XML nodes on Python 3.5 and earlier.

The flip side is what I believe is a clearer, more declarative use when creating new SVD elements.

The alternative I see is to introduce a new helper function and switch purely to *args:
```
return svd_node('peripheral',
        ET.Comment('end of %s' % module['name']),
        text_node('name', module['name']),
        text_node('baseAddress', 'module['base_addr']),
        *generate_all_interrupts(ip.get('interrupt_list')),
        svd_node('registers', *generate_all_registers(ip['registers'])))
```

To my eye this clutters up the uses and isn't as clear, making it worth the tradeoff. I'll freely admit I'm biased towards my solution, but would absolutely change this if desired.

Reply:
I've removed *args from `svd_node()`, leaving just the **kwargs text nodes.

Unfortunately, with Python 3.5 this still results in randomized ordering of the text nodes. Such SVD files are still semantically correct, and the the reordering doesn't significantly impact legibility. It also appears to have no impact on the results of `svd2rust`.